DOC: prep 1.9.0 release notes (#16248)
scipy/scipy: doc/release/1.9.0-notes.rst

Highlights of this release
**************************

- Added `scipy.optimize.milp`, new function for mixed-integer linear
  programming.
- Added `scipy.stats.fit` for fitting discrete and continuous distributions
  to data.
- Added new spline based interpolation methods for

>   `scipy.interpolate.RegularGridInterpolator`.

Reviewer: Maybe rephrase to read "Tensor-product spline interpolation modes were added to `scipy.interpolate.RegularGridInterpolator`" or some such.

- A new global optimizer (DIviding RECTangles algorithm)

>   `scipy.optimize.direct`

Reviewer 2: I have the Meson release note changes in the PR to change to it by default - but not a highlight. Would be good to add that here indeed.

************
New features
************

Lazy access to subpackages

> ==========================

Reviewer 2: I'd say that this doesn't belong here, it should go in the `Other changes` section, like the build system changes. `New features` is for subpackages only.

Before this release, all subpackages of SciPy (`cluster`, `fft`, `ndimage`,
etc.) had to be explicitly imported. Now, these subpackages are lazily loaded

> as soon as they are accessed, so that the following is possible:

Reviewer 2: add "(if desired for interactive use, it's not actually recommended for code, see `:ref:`scipy-api`).

`import scipy as sp; sp.fft.dct([1, 2, 3])`. Advantages include: making it
easier to navigate SciPy in interactive terminals, reducing subpackage import
conflicts (which before required
`import networkx.linalg as nla; import scipy.linalg as sla`),
and avoiding repeatedly having to update imports during teaching &
experimentation. Also see
[the related community specification document](https://scientific-python.org/specs/spec-0001/).

`scipy.cluster` improvements
============================

`scipy.interpolate` improvements
================================
- Speed up the ``RBFInterpolator`` evaluation with high dimensional
interpolants.
- Added new spline based interpolation methods for
`scipy.interpolate.RegularGridInterpolator` and its tutorial.
- `scipy.interpolate.RegularGridInterpolator` and `scipy.interpolate.interpn`
now accept descending ordered points.

> - ``RegularGridInterpolator`` now handles length-1 grid axes

Reviewer: Now that gh-16220 is in, let's add

- `BivariateSpline`s have grown a new method `partial_derivative` which constructs a new spline object representing a derivative of an original spline. This mirrors the corresponding functionality for univariate splines, `splder` and `BSpline.derivative`, and can substantially speed up repeated evaluation of derivatives.

`scipy.linalg` improvements
===========================
- `scipy.linalg.expm` now accepts nD arrays. It's speed is also improved.
- Minimum required LAPACK version is bumped to ``3.7.1``

`scipy.ndimage` improvements
============================

`scipy.fft` improvements

> ============================

Reviewer 2: minor: underline length needs to match title length

  additional solution information.
- Add a `integrality` parameter to `scipy.optimize.differential_evolution`,
  enabling integer constraints on parameters.
- Add a `vectorized` parameter to call a vectorized objective function only
  once per iteration. This can improve minimization speed by reducing
  interpreter overhead from the multiple objective function calls.
- The default method of ``scipy.optimize.linprog`` is now ``'highs'``.
- Added `scipy.optimize.milp`, new function for mixed-integer linear
  programming.
- added Newton-TFQMR method to ``newton_krylov``
- added support for the ``Bounds`` class in ``shgo`` and ``dual_annealing`` for
  a more uniform API across `scipy.optimize`
- added the ``vectorized`` keyword to ``differential_evolution``
- ``approx_fprime`` now works with vector-valued functions

`scipy.signal` improvements
===========================
- The new window function `scipy.signal.windows.kaiser_bessel_derived` was
  added to compute the Kaiser-Bessel derived window.
- single-precision ``hilbert`` operations are now faster as a result of more
  consistent ``dtype`` handling

`scipy.sparse` improvements
===========================

> - Add a `copy` parameter to `scipy.sparce.csgraph.laplacian`. Using inplace

Reviewer 2: formatting: double backticks for `copy`

Reviewer 2: and same for `copy=False` below

- added the ``vectorized`` keyword to ``differential_evolution``
- ``approx_fprime`` now works with vector-valued functions

`scipy.signal` improvements
===========================
- The new window function `scipy.signal.windows.kaiser_bessel_derived` was
  added to compute the Kaiser-Bessel derived window.
- single-precision ``hilbert`` operations are now faster as a result of more
  consistent ``dtype`` handling

`scipy.sparse` improvements
===========================
- Add a ``copy`` parameter to `scipy.sparce.csgraph.laplacian`. Using inplace
  computation with ``copy=False`` reduces the memory footprint.
- Add a ``dtype`` parameter to `scipy.sparce.csgraph.laplacian` for type casting.
- Add a ``symmetrized`` parameter to `scipy.sparce.csgraph.laplacian` to produce
  symmetric Laplacian for directed graphs.
- Add a ``form`` parameter to `scipy.sparce.csgraph.laplacian` taking one of the
  three values: ``array``, or ``function``, or ``lo`` determining the format of
  the output Laplacian:
  * ``array`` is a numpy array (backward compatible default);
  * ``function`` is a pointer to a lambda-function evaluating the
  Laplacian-vector or Laplacian-matrix product;

>   * ``lo`` results in the format of the `LinearOperator`.

Reviewer 3: ```suggestion
  * ``lo`` results in the format of the ``LinearOperator``.
```

`scipy.sparse.linalg` improvements
==================================
- ``lobpcg`` performance improvements for small input cases

`scipy.spatial` improvements
============================

> - Add a `order` parameter to `scipy.spatial.transform.Rotation.from_quat` and
>   `scipy.spatial.transform.Rotation.as_quat` to specify quaternion format.

Reviewer 3: ```suggestion
- Add an ``order` `parameter to `scipy.spatial.transform.Rotation.from_quat` 
  and `scipy.spatial.transform.Rotation.as_quat` to specify quaternion format.
```

`scipy.stats` improvements
==========================
- `scipy.stats.monte_carlo_test` performs one-sample Monte Carlo hypothesis
  tests to assess whether a sample was drawn from a given distribution. Besides
  reproducing the results of hypothesis tests like `scipy.stats.ks_1samp`,
  `scipy.stats.normaltest`, and `scipy.stats.cramervonmises` without small sample
  size limitations, it makes it possible to perform similar tests using arbitrary
  statistics and distributions.

> - Several ``scipy.stats`` functions support new ``axis`` (integer or tuple of

Reviewer 3: ```suggestion
- Several `scipy.stats` functions support new ``axis`` (integer or tuple of
```

  integers) and ``nan_policy`` ('raise', 'omit', or 'propagate'), and
  ``keepdims`` arguments.
  These functions also support masked arrays as inputs, even if they do not have

>   a ``scipy.stats.mstats`` counterpart. Edge cases for multidimensional arrays,

Reviewer 3: ```suggestion
  a `scipy.stats.mstats` counterpart. Edge cases for multidimensional arrays,
```

  trade off between runtime and accuracy for both the default and FFT methods of
  pdf and cdf calculation. More information can be found in the documentation
  here (to be linked).

- Added `scipy.stats.fit` for fitting discrete and continuous distributions to
  data.

- The methods ``"pearson"`` and ``"tippet"`` from `scipy.stats.combine_pvalues`
  have been fixed to return the correct p-values, resolving
  [#15373](https://github.com/scipy/scipy/issues/15373). In addition, the
  documentation for `scipy.stats.combine_pvalues` has been expanded and improved.

- Unlike other reduction functions, ``stats.mode`` didn't consume the axis
  being operated on and failed for negative axis inputs. Both the bugs have been
  fixed. Note that ``stats.mode`` will now consume the input axis and return an
  ndarray with the ``axis`` dimension removed.

- Replaced implementation of `scipy.stats.ncf` with the implementation from
  Boost for improved reliability.

- Add a `bits` parameter to `scipy.stats.qmc.Sobol`. It allows to use from 0
  to 64 bits to compute the sequence. Default is ``None`` which corresponds to
  30 for backward compatibility. Using a higher value allow to sample more

>   points. Note: it does not correspond to the output type.

Reviewer 3: ```suggestion
  points. Note: ``bits`` does not affect the output dtype.
```

- added ``qrvs`` method to ``NumericalInversePolynomial`` to match the
  situation for ``NumericalInverseHermite``

- faster random variate generation for ``gennorm`` and ``nakagami``

- ``lloyd_centroidal_voronoi_tessellation`` has been added to allow improved
  sample distributions via iterative application of Voronoi diagrams and
  centering operations

- Add `scipy.stats.qmc.PoissonDisk` to sample using the Poisson disk sampling
  method. It guarantees that samples are separated from each other by a
  given ``radius``.

- Add `scipy.stats.pmean` to calculate the weighted power mean also called
  generalized mean.

*******************
Deprecated features
*******************

- Due to collision with the shape parameter ``n`` of several distributions,
  use of the distribution ``moment`` method with keyword argument ``n`` is
  deprecated. Keyword ``n`` is replaced with keyword ``order``. 
- Similarly, use of the distribution ``interval`` method with keyword arguments
  ``alpha`` is deprecated. Keyword ``alpha`` is replaced with keyword
  ``confidence``.
- The ``'simplex'``, ``'revised simplex'``, and ``'interior-point'`` methods

>   of `scipy.optimize.linprog` are deprecated. Methods ``highs``, ``highs-ds``,

Reviewer 3: ```suggestion
  of ``scipy.optimize.linprog`` are deprecated. Methods ``highs``, ``highs-ds``,
```
I'm not sure if this will link properly.

  or ``highs-ipm`` should be used in new code.
- Support for non-numeric arrays has been deprecated from ``stats.mode``.
  ``pandas.DataFrame.mode`` can be used instead.

> - The function ``spatial.distance.kulsinski`` has been deprecated in favor
>   of ``spatial.distance.kulczynski1``.

Reviewer 3: ```suggestion
- The function `spatial.distance.kulsinski` has been deprecated in favor
  of `spatial.distance.kulczynski1`.
```

- The ``maxiter`` keyword of the truncated Newton (TNC) algorithm has been
  deprecated in favour of ``maxfun``.
- The ``vertices`` keyword of ``Delauney.qhull`` now raises a
  DeprecationWarning, after having been deprecated in documentation only
  for a long time.
- The ``extradoc`` keyword of ``rv_continuous``, ``rv_discrete`` and
  ``rv_sample`` now raises a DeprecationWarning, after having been deprecated in
  documentation only for a long time.

********************
Expired Deprecations
********************
There is an ongoing effort to follow through on long-standing deprecations.
The following previously deprecated features are affected:

- Object arrays in sparse matrices now raise an error
- Inexact indices into sparse matrices now raise an error

> - Passing ``radius=None`` to ``spatial.SphericalVoronoi`` now raises an error
>   (not adding ``radius`` defaults to 1, as before)

Reviewer 3: ```suggestion
- Passing ``radius=None`` to `scipy.spatial.SphericalVoronoi` now raises an
  error (not adding ``radius`` defaults to 1, as before)
```

- Several BSpline methods now raise an error if inputs have ``ndim > 1``.
- The ``_rvs`` method of statistical distributions now requires a ``size``
  parameter
- Passing a ``fillvalue`` that cannot be cast to the output type in

>   ``signal.convolve2d`` now raises an error

Reviewer 3: ```suggestion
  `scipy.signal.convolve2d` now raises an error
```

> - The ``spatial.distance`` now enforce that the input vectors are

Reviewer 3: ```suggestion
- `scipy.spatial.distance` now enforces that the input vectors are
```

- Removed ``cluster.vq.py_vq2``
- Removed keyword arguments ``ftol`` and ``xtol`` from
  ``optimize.minimize(method='Nelder-Mead')``
- Removed ``signal.windows.hanning``
- Removed LAPACK ``gegv`` functions from ``linalg``; this raises the minimally
  required LAPACK version to 3.7.1
- Removed ``spatial.distance.matching``
- Removed the alias ``scipy.random`` for ``numpy.random``
- Removed docstring related functions from ``scipy.misc`` (``docformat``,
  ``inherit_docstring_from``, ``extend_notes_in_docstring``,
  ``replace_notes_in_docstring``, ``indentcount_lines``, ``filldoc``,
  ``unindent_dict``, ``unindent_string``).
- Removed ``linalg.pinv2``

******************************
Backwards incompatible changes
******************************

- Several `scipy.stats` functions now convert ``np.matrix`` to ``np.ndarray``s
  before the calculation is performed. In this case, the output will be a scalar
  or ``np.ndarray`` of appropriate shape rather than a 2D ``np.matrix``.
  Similarly, while masked elements of masked arrays are still ignored, the
  output will be a scalar or ``np.ndarray`` rather than a masked array with
  ``mask=False``.

> - The default method of ``scipy.optimize.linprog`` is now ``'highs'``, not

Reviewer 3: ```suggestion
- The default method of `scipy.optimize.linprog` is now ``'highs'``, not
```

  generalized mean.

*******************
Deprecated features
*******************

- Due to collision with the shape parameter ``n`` of several distributions,
  use of the distribution ``moment`` method with keyword argument ``n`` is
  deprecated. Keyword ``n`` is replaced with keyword ``order``. 
- Similarly, use of the distribution ``interval`` method with keyword arguments
  ``alpha`` is deprecated. Keyword ``alpha`` is replaced with keyword
  ``confidence``.
- The ``'simplex'``, ``'revised simplex'``, and ``'interior-point'`` methods
  of `scipy.optimize.linprog` are deprecated. Methods ``highs``, ``highs-ds``,
  or ``highs-ipm`` should be used in new code.
- Support for non-numeric arrays has been deprecated from ``stats.mode``.
  ``pandas.DataFrame.mode`` can be used instead.
- The ``extradoc`` keyword of ``rv_continuous``, ``rv_discrete`` and
  ``rv_sample`` now raises a DeprecationWarning, after having been deprecated in
  documentation only for a long time already.

********************
Expired Deprecations
********************
There is an ongoing effort to follow through on long-standing deprecations.
The following previously deprecated features are affected:

> - Object arrays in sparse matrices now raise an error

Reviewer 4: You need a blank line before the beginning of the bulleted list.  (I was going to link to the comment about this in the "missing bits" document, but apparently the link to that document in the rendered docs was broken by some recent reshuffling of the docs files. I see if I can fix that and submit a PR.)

Reviewer 4: Here's the link to the relevant section of the docs from a PR that fixes the reference to the "missing bits" file: [Bulleted lists](https://output.circle-artifacts.com/output/job/7265df6e-f1f4-4e37-9ca6-2cd6a7b903f1/artifacts/0/html-scipyorg/dev/contributor/missing-bits.html#bulleted-lists)

  ``mask=False``.
- The default method of ``scipy.optimize.linprog`` is now ``'highs'``, not
  ``'interior-point'`` (which is now deprecated), so callback functions and some
  options are no longer supported with the default method.
- For `scipy.stats.combine_pvalues`, the sign of the test statistic returned
  for the method ``"pearson"`` has been flipped so that higher values of the
  statistic now correspond to lower p-values, making the statistic more
  consistent with those of the other methods and with the majority of the
  literature.
- `scipy.linalg.expm` due to historical reasons was using the sparse
  implementation and thus was accepting sparse arrays. Now it only works with
  nDarrays. For sparse usage, `scipy.sparse.linalg.expm` needs to be used
  explicitly.
- The definition of `scipy.stats.circvar` has reverted to the one that is
  standard in the literature; note that this is not the same as the square of
  `scipy.stats.circstd`.
- Remove inheritance to `QMCEngine` in `MultinomialQMC` and
  `MultivariateNormalQMC`. It removes the methods `fast_forward` and `reset`.
- Init of `MultinomialQMC` now require the number of trials with `n_trials`.

>   Hence, `MultinomialQMC.random` output has now the correct shape ``(n, pvals)``.

Reviewer 3: ```suggestion
  Hence, `MultinomialQMC.random` output has now the correct shape ``(n, pvals)``.
- Several function-specific warnings (``F_onewayConstantInputWarning``,
  ``F_onewayBadInputSizesWarning``, ``PearsonRConstantInputWarning``, 
  ``PearsonRNearConstantInputWarning``, ``SpearmanRConstantInputWarning``, and
  ``BootstrapDegenerateDistributionWarning``) have been replaced with more
  general warnings.
```

to use the approach of Wang and Zhang from their 2008 conference paper
*Simpson’s rule based FFT method to compute densities of stable distribution*,
making this method more competitive with the default method. In addition,
users now have the option to change the parametrization of the Levy Stable
distribution to Nolan's "S0" parametrization which is used internally by
SciPy's pdf and cdf implementations. The "S0"  parametrization is described in
Nolan's paper [*Numerical calculation of stable densities and distribution
functions*](https://doi.org/10.1080/15326349708807450) upon which SciPy's
implementation is based. "S0" has the advantage that ``delta`` and ``gamma``
are proper location and scale parameters. With ``delta`` and ``gamma`` fixed,
the location and scale of the resulting distribution remain unchanged as
``alpha`` and ``beta`` change. This is not the case for the default "S1"
parametrization. Finally, more options have been exposed to allow users to
trade off between runtime and accuracy for both the default and FFT methods of
pdf and cdf calculation. More information can be found in the documentation
here (to be linked).

> - Added `scipy.stats.fit` for fitting discrete and continuous distributions to

Reviewer 2: Just noticing this now, `fit` is a _very_ generic name. I can't remember it being proposed/discussed. Will comment on gh-15436.

data.

- The methods ``"pearson"`` and ``"tippet"`` from `scipy.stats.combine_pvalues`
have been fixed to return the correct p-values, resolving
[#15373](https://github.com/scipy/scipy/issues/15373). In addition, the
documentation for `scipy.stats.combine_pvalues` has been expanded and improved.

- Unlike other reduction functions, ``stats.mode`` didn't consume the axis
being operated on and failed for negative axis inputs. Both the bugs have been
fixed. Note that ``stats.mode`` will now consume the input axis and return an
ndarray with the ``axis`` dimension removed.

- Replaced implementation of `scipy.stats.ncf` with the implementation from

> Boost.

Reviewer 3: ```suggestion
Boost for improved reliability.
```

- Add a `bits` parameter to `scipy.stats.qmc.Sobol`. It allows to use from 0
to 64 bits to compute the sequence. Default is ``None`` which corresponds to
30 for backward compatibility. Using a higher value allow to sample more
points. Note: it does not correspond to the output type.

- Add a `integers` method to `scipy.stats.qmc.QMCEngine`. It allows sampling
integers using any QMC sampler.

- improved the fit speed and accuracy of ``stats.pareto``

- added ``qrvs`` method to ``NumericalInversePolynomial`` to match the
situation for ``NumericalInverseHermite``

> - faster random variate generation for ``gennorm``

Reviewer 3: ```suggestion
- faster random variate generation for ``gennorm`` and ``nakagami``
```
gh-15360

- ``lloyd_centroidal_voronoi_tessellation`` has been added to allow improved
sample distributions via iterative application of Voronoi diagrams and

> centering operations
> 
> Other

Reviewer 5: ```suggestion
centering operations.

- Add `scipy.stats.qmc.PoissonDisk` to sample using the Poisson disk sampling method. It guarantees that samples are separated from each other by a given ``radius``.

Other
```

-----

A draft developer CLI is available for SciPy, leveraging the ``doit``,
``click`` and ``rich-click`` tools. For more details, see

> [gh-15959](https://github.com/scipy/scipy/pull/15959).

Author: Maybe we can fuse the two "Other" themed sections?

Reviewer 2: Yes, good idea.

*******************
Deprecated features
*******************

- Due to collision with the shape parameter ``n`` of several distributions,
use of the distribution ``moment`` method with keyword argument ``n`` is
deprecated. Keyword ``n`` is replaced with keyword ``order``. 
- Similarly, use of the distribution ``interval`` method with keyword arguments
``alpha`` is deprecated. Keyword ``alpha`` is replaced with keyword
``confidence``.
- The ``'simplex'``, ``'revised simplex'``, and ``'interior-point'`` methods
of `scipy.optimize.linprog` are deprecated. Methods ``highs``, ``highs-ds``,
or ``highs-ipm`` should be used in new code.
- Support for non-numeric arrays has been deprecated from ``stats.mode``.
``pandas.DataFrame.mode`` can be used instead.
- The ``extradoc`` keyword of ``rv_continuous``, ``rv_discrete`` and
``rv_sample`` now raises a DeprecationWarning, after having been deprecated in
documentation only for a long time already.

********************

> Expired Deprecations

Reviewer 5: @h-vetinari @j-bowhay FYI if there are other entries to add, please add them here as suggestion so that it's easier for Tyler.

Reviewer 6: I've tried to keep the wiki up-to-date with the issues in https://github.com/scipy/scipy/issues/15765 as they were closed. To the best of my knowledge, the list as below is complete.

Reviewer 5: Thanks for confirming.

``mask=False``.
- The default method of ``scipy.optimize.linprog`` is now ``'highs'``, not
``'interior-point'`` (which is now deprecated), so callback functions and some
options are no longer supported with the default method.
- For `scipy.stats.combine_pvalues`, the sign of the test statistic returned
for the method ``"pearson"`` has been flipped so that higher values of the
statistic now correspond to lower p-values, making the statistic more
consistent with those of the other methods and with the majority of the
literature.
- `scipy.linalg.expm` due to historical reasons was using the sparse
implementation and thus was accepting sparse arrays. Now it only works with
nDarrays. For sparse usage, `scipy.sparse.linalg.expm` needs to be used
explicitly.
- The definition of `scipy.stats.circvar` has reverted to the one that is
standard in the literature; note that this is not the same as the square of
`scipy.stats.circstd`.
- Remove inheritance to `QMCEngine` in `MultinomialQMC` and
`MultivariateNormalQMC`. It removes the methods `fast_forward` and `reset`.
- Init of `MultinomialQMC` now require the number of trials with `n_trials`.

> Hence, `MultinomialQMC.random` output has now the correct shape ``(n, pvals)``.

Reviewer 3: Email [here](https://mail.python.org/archives/list/scipy-dev@python.org/thread/QEMS6VZSBHTRB6OVSDBULJNNOE7PFAOV/#QEMS6VZSBHTRB6OVSDBULJNNOE7PFAOV), discussion [here](https://github.com/scipy/scipy/pull/15945). This was definitely a bug.

*************
Other changes
*************

- The SciPy contributor guide has been reorganized and updated
(see [#15947](https://github.com/scipy/scipy/pull/15947) for details).

- QUADPACK Fortran routines in `scipy.integrate`, which power
`scipy.integrate.quad`, have been marked as `recursive`. This should fix rare
issues in multivariate integration (`nquad` and friends) and obviate the need
for compiler-specific compile flags (`/recursive` for ifort etc). Please file

> an issue if this change turns out problematic for you.

Reviewer: Also FITPACK routines in `scipy.interpolate`, which power `splrep`, `splev` etc, and `*UnivariateSpline` and `*BivariateSpline` classes.

* `#16031 <https://github.com/scipy/scipy/issues/16031>`__: BUG: scipy.sparse.linalg.norm does not work on sparse arrays
* `#16036 <https://github.com/scipy/scipy/issues/16036>`__: Missing \`f\` prefix on f-strings
* `#16054 <https://github.com/scipy/scipy/issues/16054>`__: Bug: Meson build with dev.py fails to detect SciPy with debian...
* `#16065 <https://github.com/scipy/scipy/issues/16065>`__: BUG: Gitpod build with \`python runtests.py\` fails; move to...
* `#16074 <https://github.com/scipy/scipy/issues/16074>`__: BUG: refguide check fails with \`numpydoc==1.3\`
* `#16081 <https://github.com/scipy/scipy/issues/16081>`__: CI, MAINT: minor refguide failure with stats.describe
* `#16121 <https://github.com/scipy/scipy/issues/16121>`__: DOC: scipy.interpolate.RegularGridInterpolator and interpn works...
* `#16162 <https://github.com/scipy/scipy/issues/16162>`__: BUG: curve_fit gives wrong results with Pandas float32
* `#16171 <https://github.com/scipy/scipy/issues/16171>`__: BUG: scipy.stats.multivariate_hypergeom.rvs raises ValueError...

***********************
Pull requests for 1.9.0
***********************

> * `#354 <https://github.com/scipy/scipy/pull/354>`__: Faster expm

Reviewer 3: ```suggestion
```
This wasn't actually merged; it was superseded. Perhaps the code that generates this list could use a little tweak? Not too important, though.

* Patrick N. Raanes (1) +
* Tyler Reddy (93)
* Pamphile Roy (192)
* Vivek Roy (2) +
* Niyas Sait (2) +
* Atsushi Sakai (25)
* Mazen Sayed (1) +
* Eduardo Schettino (5) +
* Daniel Schmitz (6) +
* Eli Schwartz (3) +
* SELEE (2) +
* Namami Shanker (4)
* siddhantwahal (1) +
* Gagandeep Singh (8)
* Soph (1) +
* Shivnaren Srinivasan (1) +
* Scott Staniewicz (1) +
* Leo C. Stein (4)
* Albert Steppi (7)
* Christopher Strickland (1) +
* Kai Striega (4)
* Søren Fuglede Jørgensen (1)
* Aleksandr Tagilov (1) +
* Masayuki Takagi (1) +
* Sai Teja (1) +
* Ewout ter Hoeven (1) +
* Will Tirone (2)
* Bas van Beek (7)
* Dhruv Vats (1)
* H. Vetinari (4)
* Arthur Volant (1)
* Samuel Wallan (5)
* Stefan van der Walt (8)
* Warren Weckesser (81)
* Anreas Weh (1)
* Nils Werner (1)
* Aviv Yaish (1) +
* Dowon Yi (1)
* Rory Yorke (1)
* Yosshi999 (1) +
* yuanx749 (2) +
* Gang Zhao (23)
* ZhihuiChen0903 (1)
* Pavel Zun (1) +
* David Zwicker (1) +

A total of 152 people contributed to this release.
People with a "+" by their names contributed a patch for the first time.
This list of names is automatically generated, and may not be fully complete.

***********************
Issues closed for 1.9.0

> ***********************

Reviewer 2: Indeed, see `tools/gh_lists.py`.

Grabbing issues based on date of closure isn't going to work, because a lot are closed because fixed in older releases, invalid reports, etc.

As of now, you should set milestones on issues that you close. A way to improve the script could be to parse commit messages for "Closes xxx", but even that isn't going to be foolproof since I often see that people fail to add that info to commit messages and only add it to PR descriptions.

> ***********************

Reviewer: A possible automation could be to auto-add a milestone to an issue which is closed by a PR  with a milestone. Maybe that's a feature request to github or maybe it's possible to do with their api.

> ***********************

Reviewer 3: I started to do this for issues I was involved with. I thought I should make sure I'm doing the right thing before going further. For instance, I added the milestone to gh-14622 because it was a legitimate issue that was closed by a PR, whereas I did not add a milestone to gh-15160 because it was reported due to a misunderstanding. LMK if I should keep going this way.

> ***********************

Reviewer 2: Yes, that sounds right, thanks Matt!

> ***********************

Reviewer 3: OK, done. I added a lot of milestones. 
I accidentally created a 1.90 milestone in the process. Do you know how to remove that from the list?

> ***********************

Reviewer 2: Thanks Matt. I deleted the 1.90 milestone.
